clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java

```diff
@@ -0,0 +1,324 @@
+        maybeTransitionToFailureState(responseError);
+
+        if (responseError == Errors.GROUP_AUTHORIZATION_FAILED) {
+            GroupAuthorizationException error = GroupAuthorizationException.forGroupId(membershipManager.groupId());
+            logger.error("GroupHeartbeatRequest failed due to group authorization failure: {}", error.getMessage());
+            nonRetriableErrorHandler.handle(error);
+        } else if (responseError == Errors.INVALID_REQUEST) {
+            logger.error("GroupHeartbeatRequest failed due to invalid request error: {}",
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.INVALID_REQUEST.exception());
+        } else if (responseError == Errors.GROUP_MAX_SIZE_REACHED) {
+            logger.error("GroupHeartbeatRequest failed due to the max group size limit: {}",
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(responseError.exception());
+        } else if (responseError == Errors.UNSUPPORTED_ASSIGNOR) {
+            logger.error("GroupHeartbeatRequest failed due to unsupported assignor {}: {}",
+                membershipManager.assignorSelection(), response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.UNSUPPORTED_ASSIGNOR.exception());
+        } else if (responseError == Errors.UNRELEASED_INSTANCE_ID) {
+            logger.error("GroupHeartbeatRequest failed due to the instance id {} was not released: {}",
+                membershipManager.groupInstanceId().orElse("null"),
+                response.data().errorMessage());
+            nonRetriableErrorHandler.handle(Errors.UNRELEASED_INSTANCE_ID.exception());
+        } else if (responseError == Errors.FENCED_MEMBER_EPOCH ||
+                responseError == Errors.UNKNOWN_MEMBER_ID) {
+            membershipManager.fenceMember();
+        } else {
+            // If the manager receives an unknown error - there could be a bug in the code or a new error code
+            logger.error("GroupHeartbeatRequest failed due to unexpected error: {}", responseError);
+            membershipManager.transitionToFailure();
+        }
+    }
+
+    private void maybeTransitionToFailureState(Errors error) {
+        if (fatalErrors.contains(error)) {
+            membershipManager.transitionToFailure();
+        }
+    }
+
+    private void logInfo(final String message,
+                         final ConsumerGroupHeartbeatResponse response,
+                         final long currentTimeMs) {
+        logger.info("{} in {}ms: {}",
+            message,
+            heartbeatRequestState.remainingBackoffMs(currentTimeMs),
+            response.data().errorMessage());
+    }
+
+    static class HeartbeatRequestState extends RequestState {
+        long heartbeatIntervalMs;
+        final Timer heartbeatTimer;
+        // TODO: Add an inflight request tracker and test sending multiple heartbeat concurrently
```

Comment:
Just for my understanding, what's the idea behind this TODO? I thought we had inflight req handling in the parent `RequestState`, that already identifies `log.trace("An inflight request already exists for {}", this);`. And what would be the concurrent scenario if there is a single background thread sending heartbeats and we are not resending while there is one inflight?

Reply:
i think it was a note from before, removed.